forge/src/main/java/me/lucko/luckperms/forge/ForgeClassPathAppender.java

```diff
@@ -0,0 +1,135 @@
+ *
+ *  Copyright (c) lucko (Luck) <luck@lucko.me>
+ *  Copyright (c) contributors
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ */
+
+package me.lucko.luckperms.forge;
+
+import cpw.mods.cl.ModuleClassLoader;
+import cpw.mods.modlauncher.Launcher;
+import cpw.mods.modlauncher.TransformingClassLoader;
+import me.lucko.luckperms.common.plugin.classpath.ClassPathAppender;
+import me.lucko.luckperms.common.plugin.logging.PluginLogger;
+import sun.misc.Unsafe;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+
+public class ForgeClassPathAppender extends URLClassLoader implements ClassPathAppender {
+    private final PluginLogger logger;
+    private final Set<String> invalidClasses;
+    private final Map<String, ClassLoader> parentLoaders;
+
+    public ForgeClassPathAppender(PluginLogger logger) {
+        super(new URL[0], (ModuleClassLoader) Launcher.class.getClassLoader());
+        this.logger = logger;
+        this.invalidClasses = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        this.parentLoaders = getParentLoaders((TransformingClassLoader) Thread.currentThread().getContextClassLoader());
+    }
+
+    @Override
+    protected Class<?> findClass(String name) throws ClassNotFoundException {
+        if (invalidClasses.contains(name)) {
+            throw new ClassNotFoundException(name);
+        }
+
+        invalidClasses.add(name);
+        Class<?> validClass = super.findClass(name);
+        invalidClasses.remove(name);
+        return validClass;
+    }
+
+    @Override
+    public void addJarToClasspath(Path file) {
```

Comment:
Some comments to explain what's going on in this method would be good!

Reply:
Comments added